### ISSUE TITLE
PRNG random functions use wrong index

### ISSUE THREAD

**Issue by Cosine256** [spelunky-fyi/overlunky#302]
The PRNG functions `random_chance`, `random_float`, `random_index`, and `random_int` do not adjust the `type` argument from the `PRNG_CLASS` Lua range (1-10) to the C++ range (0-9). As a result, they always poll the wrong index. The functions `get_pair` and `set_pair` do not have this problem. The random functions also have no bounds checks on the type, unlike `get_pair` and `set_pair`.

**Comment by Dregu** [comment]
Yeah it's pretty stupid, but the random functions should actually work correctly. There is no "lua range" for the enum, it's same on both sides.

About fixing this discrepancy though... On one hand the get_pair function "index" is ambiguous so I wouldn't feel bad just changing the range to current PRNG_CLASS. On the other hand, the random functions are documented to use the enum, which we are allowed to change on a whim and it's the users fault for hardcoding numbers where enum is expected... Don't really know which one to fix.

**Comment by Cosine256** [comment]
I looked at this again and I see what you mean now. 

I originally encountered this problem in a script where I'm trying to read and display the state of the 10 PRNG pairs in a UI. I needed to iterate through them and I determined that 1-10 was the range that got them all. I labelled them by matching their 1-10 index with PRNG_CLASS (if they had a match) without looking into the actual values in that enum. I hadn't noticed that PROCEDURAL_SPAWNS (0) wasn't a label in my UI at all since it's not within the 1-10 range. Later, I tried to use the `random_float` function on ENTITY_VARIATION (3) to advance the RNG and I noticed it was advancing the next index in my UI, which is when I reported this issue. ENTITY_VARIATION (3) was being polled by things like teleporting and jiangshi jump trajectories, so I assumed my UI was correct.

It seems the index I was watching was actually PARTICLES (2) and my UI was wrong. I wish we knew more about what these PRNG calls actually do. It makes sense that the teleport particles would use that index, but the teleport distance also uses it, as well as the jiangshi, and neither use ENTITY_VARIATION (3), while mattocking a tile uses both.

In a world where we didn't care about breaking changes, I'd rather see `get_pair` and `set_pair` be changed to accept PRNG_CLASS instead of "index". Changing the enum and random functions instead is probably less harmful, but that change would still break any scripts that had to hard-code an index that isn't in the enum.
